app/src/main/java/com/babylon/wallet/android/data/repository/tokenprice/TokenPriceRepository.kt

```diff
@@ -0,0 +1,113 @@
+        runCatching {
+            val tokensPrices = tokenPriceApi.tokens()
+                .toResult(
+                    mapError = { mapTokenPriceApiError(this) }
+                )
+                .getOrElse {
+                    Timber.e("failed to fetch tokens prices with exception: ${it.message}")
+                    return@withContext Result.failure(it)
+                }
+
+            tokenPriceDao.insertTokensPrice(
+                tokensPrices = tokensPrices.asEntity()
+            )
+        }
+    }
+
+    override suspend fun getTokensPrices(
+        resourcesAddresses: Set<String>,
+        lsusAddresses: Set<String>
+    ): Result<List<TokenPrice>> = withContext(ioDispatcher) {
+        runCatching {
+            val allResourcesAddresses = resourcesAddresses + lsusAddresses
+
+            val tokensPrices = tokenPriceDao.getTokensPrices(
+                addresses = allResourcesAddresses,
+                minValidity = tokenPriceCacheValidity()
+            )
+
+            val remainingResourcesAddresses = resourcesAddresses subtract tokensPrices.map { it.resourceAddress }.toSet()
+            val remainingLsusAddresses = lsusAddresses subtract tokensPrices.map { it.resourceAddress }.toSet()
+
+            if (remainingResourcesAddresses.isNotEmpty() || remainingLsusAddresses.isNotEmpty()) {
+                tokenPriceApi.priceTokens(
+                    tokensAndLsusPricesRequest = TokensAndLsusPricesRequest(
+                        currency = "USD",
+                        lsus = remainingLsusAddresses.toList(),
+                        tokens = remainingResourcesAddresses.toList()
+                    )
+                ).toResult(
+                    mapError = { mapTokenPriceApiError(this) }
+                ).mapCatching { currentTokensAndLsusPrices ->
+                    // update the database with the new tokens prices
+                    tokenPriceDao.insertTokensPrice(tokensPrices = currentTokensAndLsusPrices.asEntity())
+                    // return all tokens prices from database
+                    tokenPriceDao.getTokensPrices(
+                        addresses = allResourcesAddresses,
+                        minValidity = tokenPriceCacheValidity()
+                    ).toTokenPrice()
+                }.getOrElse {
+                    Timber.e("failed to fetch tokens prices with exception: ${it.message}")
+                    emptyList()
+                }
+            } else {
+                tokensPrices.toTokenPrice()
+            }
+        }
+    }
+
+    private fun mapTokenPriceApiError(responseBody: ResponseBody?): RadixWalletException.GatewayException {
+        val error = Serializer.kotlinxSerializationJson.decodeFromString<TokensPricesErrorResponse>(
+            responseBody?.string().orEmpty()
+        )
+        return RadixWalletException.GatewayException.HttpError(code = 500, message = error.detail ?: "no message")
```

Comment:
Why always 500?

Reply:
because the token price service doesn't return an Http error code.
Any idea how would I handle this? 👀